drivers/dcf77/dcf77.c

```diff
@@ -0,0 +1,436 @@
+ *
+ * @author      Michel Gerlach <michel.gerlach@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+#include <string.h>
+
+#include "log.h"
+#include "assert.h"
+#include "xtimer.h"
+#include "timex.h"
+#include "periph/gpio.h"
+
+#include "dcf77.h"
+#include "dcf77_internal.h"
+#include "dcf77_params.h"
+
+
+#define ENABLE_DEBUG    (1)
+#include "debug.h"
+
+
+/* Every pulse send by the DCF longer than 200ms new Minute is beginning */
+#define PULSE_START_HIGH_THRESHOLD       (1500000)  /*~1000ms*/
+/* Every pulse send by the DCF longer than 100ms is interpreted as 1 */
+#define PULSE_WIDTH_THRESHOLD       (140000U)       /*~150ms*/
+/* If an expected pulse is not detected within 1s, something is wrong */
+#define TIMEOUT                     (2500000)       /*~0200ms*/
+/* The DCF sensor cannot measure more than once a second */
+#define DATA_HOLD_TIME              (US_PER_SEC)
+
+#define READING_CYCLE               59
+
+
+/**
+ * @brief   Wait until the pin @p pin has level @p expect
+ *
+ * @param   pin     GPIO pin to wait for
+ * @param   expect  Wait until @p pin has this logic level
+ * @param   timeout Timeout in µs
+ *
+ * @retval  0       Success
+ * @retval  -1      Timeout occurred before level was reached
+ */
+static inline int _wait_for_level(gpio_t pin, bool expect, unsigned timeout)
+{
+    while ((gpio_read(pin) != expect) && (timeout != 0)) {
+        xtimer_usleep(1);
+        timeout--;
+    }
+
+    return (timeout > 0) ? 0 : -1;
+}
+
+/**
+ * @brief   get the bit of a single second
+ *
+ *
+ * @param[in]  dev        Device Params of DCF77
+ * @param[in]  startBit   first binary value
+ * @param[in]  endBit     expected binary value
+ * @param[in]  threshold  threshold to pass
+ *
+ * @retval `DCF77_OK`         Success
+ * @retval `DCF77_TIMEOUT`    Reading data timed out (bad receiving)
+ */
+static uint8_t _read(bool *dest, gpio_t pin, bool startBit, bool endBit,
+                     uint32_t threshold)
+{
+
+    uint32_t start = 0, end = 0;
+
+    if (_wait_for_level(pin, startBit, TIMEOUT)) {
+        DEBUG("Timeout wait for %d\n", startBit);
+        return DCF77_TIMEOUT;
+    }
+    start = xtimer_now_usec();
+
+    if (_wait_for_level(pin, endBit, TIMEOUT)) {
+        DEBUG("Timeout wait for %d\n", endBit);
+        return DCF77_TIMEOUT;
+
+    }
+    end = xtimer_now_usec();
+
+    if ((end - start) > threshold) {
+        *dest = 1;
+    }
+    else {
+        *dest = 0;
+    }
+    return DCF77_OK;
+}
+
+/**
+ * @brief   Decode the current Bit and save the values
+ *
+ *
+ * @param[in]  bitNumber  Current Bit
+ * @param[in]  result     value of the Current Bit
+ * @param[in]  *data      Datastruct of the DCF77
+ *
+ * @retval `DCF_TIMEOUT`    Reading data timed out
+ * @retval `DCF77_NOCSUM`   Parity Error
+ * @retval `DCF77_OK`       Success
+ */
+
+static int8_t _decodeInformation(uint8_t bitNumber, uint8_t result,
+                                 dcf77_data_t *data)
+{
+    switch (bitNumber) {
```

Comment:
There is a lot of duplication in the switch cases. Maybe this can be reduced a little bit. One idea could be using fall through, or adding an additional parameter to the function type? Just think of it for a moment.